peps/pep-0807.rst

```diff
@@ -0,0 +1,398 @@
+----------------------------
+
+All Python package uploading is currently "endpoint driven," in the sense
+uploading clients (like twine and uv) are given an upload URL (and
+**not** merely a domain name).
+
+For example, to upload to PyPI, uploading clients are expected to connect
+to ``https://upload.pypi.org/legacy/``.
+
+The discovery mechanism proposed below takes advantage of this fact to
+allow single domains to advertise support for multiple indices
+(and their corresponding upload endpoints).
+
+The discovery mechanism is as follows:
+
+1. The uploading client is given an upload URL, e.g.
+   ``https://upload.example.com/legacy/``.
+
+2. The uploading client extracts the *path component* of the URL,
+   as defined in :rfc:`3986`. If the path component is empty,
+   the empty string should be used.
+
+   For the above example, the path component is
+   ``/legacy/``.
+
+3. The uploading client takes the SHA2-256 hash of the path component,
+   producing the *discovery key*.
+
+   For the above example, the discovery key is
+   ``af030c06750716b1b35852298fe852b90def13dcbd012a5fe5148470f1206bfc``. [#fn-hash]_
+
+4. The uploading client constructs a *discovery URL* by taking the
+   scheme and authority components (as defined in :rfc:`3986`)
+   of the upload URL and appending ``/.well-known/pytp/``
+   and the discovery key.
+
+   For the above example, the discovery URL is
+   ``https://upload.example.com/.well-known/pytp/af030c06750716b1b35852298fe852b90def13dcbd012a5fe5148470f1206bfc``.
+
+5. The uploading client performs an HTTP GET request to the discovery URL.
+
+6. The server responds with a ``200 OK`` status code and a body
+   containing a JSON object if the index supports Trusted Publishing
+   for the given upload URL. The JSON object **MUST** contain the following
+   fields:
+
+   - ``audience-endpoint``: a string containing the URL of the OIDC
+     audience endpoint to be used during token exchange.
+   - ``token-mint-endpoint``: a string containing the URL of the
+     token minting endpoint to be used during token exchange.
+
+   For the above example, a valid response body would be:
+
+   .. code-block:: json
+
+       {
+            "audience-endpoint": "https://upload.example.com/_/oidc/audience",
+            "token-mint-endpoint": "https://upload.example.com/_/oidc/mint-token"
+       }
+
+If the server does not support Trusted Publishing for the given
+upload URL, it **MUST** respond with a ``404 Not Found`` status code.
+When responding with a ``404 Not Found``, the server **SHOULD NOT**
+include a response body. If a response body is included, it **MUST**
+be ignored by the client.
+
+Servers **MAY** additionally respond with any other standard HTTP
+error code in the 400 or 500 range to indicate an error condition.
+
+Non-``200 OK``, non-``404 Not Found`` responses **MAY** include a body which,
+if present, **MUST** be a JSON object containing an
+:ref:`Error Response <error-responses>`.
```

Comment:
All titles are implicitly link targets within the same document, prefer using this over `:ref:`, which has a global namespace.
```suggestion
`Error Response`_.
```